Sql styling

arjank opened the pull request:
Styling for sql queries

Commit:
SQL / MySQL coding style standards

Signed-off-by: Arjan Kleene <arjan.kleene@dealerdirect.nl>

Commit:
SQL / MySQL coding style standards

- Updates ToC
- Adds insert ... set ... statement
- Adds formatting for php

Signed-off-by: Arjan Kleene <arjan.kleene@dealerdirect.nl>

Commit:
SQL / MySQL coding style standards

- Updates formatting
- Adds spelling entries

Signed-off-by: Arjan Kleene <arjan.kleene@dealerdirect.nl>

Potherca reviewed (changes requested):
The MySql logo image is 167px wide, not 200px like the other images.

Please replace with [a 200px version](https://upload.wikimedia.org/wikipedia/en/thumb/6/62/MySQL.svg/200px-MySQL.svg.png)

Potherca reviewed (changes requested):
It should be made clear that instances of "English" should be "American English", either by stating this at the beginning of the document or by changing all occurrences.

Potherca commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+
+## Formatting
+
+- Keywords must be written in upper case
```
Typo `must` should be capitalised `MUST`.

Potherca commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+
+The `UPDATE` keyword MUST be followed by the table name that will be affected.
+The `SET` keyword MUST be on the next line and and MUST not be indented. Column
+names should be on separate lines, followed by an equals sign and the new value
```
Typo `should` should be capitalised `SHOULD`.

Potherca commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+- Indents MUST be two spaces
+- Keywords MUST be followed by exactly one space
+- Keywords that consist of multiple words MUST be written on one line
+- Queries SHOULD be parametrized statements
```
Typo `parametrized` should be `parameterised`

arjank replied:
No, `parameterised` is the British English spelling.

Potherca replied:
My bad. I thought the "z" was the US English way.

I keep mixing those two up :disappointed:

Potherca replied:
Haha (quoting from [the wiki page on this subject](https://en.wikipedia.org/wiki/American_and_British_English_spelling_differences#-ise.2C_-ize_.28-isation.2C_-ization.29)):

> The -ize spelling is often incorrectly seen as an Americanism in Britain

I guess they're not kidding :laughing:

mjrider commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+### Limit
+
+A `LIMIT` clause SHOULD only be used together with an `ORDER BY` clause.
+
```
I disagree here
Offset should be used as a separate keyword instead of being a part of the limit statement, so limit 20,10 is invalid

Limit is valid without order by, just give me 10 something from the table, i do not care about the order/grouping
Even with offset it could be valid for paging, except it would be better to do it in a begin/end block.

arjank replied:
Makes sense, `LIMIT 10 OFFSET 20` is more readable than `LIMIT 20, 10`. I'll change this.

arjank replied:
The documentation still allows the use of limit without order by. It is a recommendation, not a requirement.

mjrider replied:
i read should as 'it is required unless you have a very good argumented corner case', which is imho a to strong statement here, but we could always change the guide later if we have to ignore this rule to often

mjrider reviewed (commented):
I think the subject 'subqueries' is missing.

Potherca commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+- Queries SHOULD be parametrized statements
+- Parameters SHOULD be specified by name
+- Operators MUST be surrounded by one space
+- Table names, column names and database names SHOULD NOT be surrounded with backticks
```
Doesn't this pose problems for columns that are named a [reserved word](https://dev.mysql.com/doc/refman/5.7/en/keywords.html)? (Or is this effect negated by using lowercase?)

I am aware backticks are not part of the ANSI SQL standard. If this is the reason to avoid them I would like that to be added...

arjank replied:
I'll add something about reserved words.

Potherca commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+Delete queries SHOULD not be used.
+
+If a delete query is required, the table from which the rows are deleted MUST be
+specified on the same line as the `DELETE FROM` keyword. Delete queries also
```
Unneeded adverb `also`.

Potherca commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+aggregated columns. Columns that are not part of the `GROUP BY` clause and not
+aggregated, SHOULD NOT appear in the field list.
+
+If aggregation functions are not used, `GROUP BY` also SHOULD NOT be used.
```
Unneeded adverb `also`.

Potherca commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+
+### Delete queries
+
+Delete queries SHOULD not be used.
```
It might be wise to offer an alternative (like adding a "deleted" column or using an archive table)

mjrider commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+
+### Delete queries
+
+Delete queries SHOULD not be used.
```
Why? Because a deleting data is good for a healthy dataset.
it is even required by law that we are able to delete data, so  making a policy which say you should not is odd
deleting data should happen with care, and should have a where statement

arjank replied:
This document still allows delete queries. It also requires a where clause if you use a delete query. I want people to think twice before writing a query that deletes data without a restore option.

mjrider replied:
If that is the argumentation, i would jus remove it, because having a way to rollback state should not be defined in a styleguide for SQL. An advise could be given, but that advise should not use the terms must or shall.

Potherca commented on code-styling/mysql/README.md:
````diff
@@ -0,0 +1,574 @@
+AND created_at > '2016-06-30 12:34:56'
+```
+
+### Joins
````
Shouldn't the "Joins" header be a h4 under "Select queries"?

arjank replied:
Possibly, but joins are also possible in other types of queries.

Potherca replied:
In that case I would suggest they belong after all query sections...

Potherca commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+columns MUST have a default value. The `id` column MUST NOT be specified in
+insert queries unless those queries are used in some kind of data replication.
+
+If an insert query specifies more than 5 columns, the `SET` syntax is
```
Typo `5` should be `five`<sup>[1]</sup>

<sup>[1]: In writing it is common to use words rather than digits for numbers under 10 (or at the beginning of a sentence).</sup>

Potherca commented on code-styling/mysql/README.md:
````diff
@@ -0,0 +1,574 @@
+WHERE id = 5;
+```
+
+### Other select criteria
````
It seems to me that all of these (Between, Grouping, Ordering, and Limit) should be underneath the "Select queries" section.

Potherca commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+Inline queries are sql queries in a file of another programming language, such
+as a PHP file with sql queries.
+
+### Queries in php files
```
I am unsure whether this belongs here or if we should add/update this to a PHP standard..

arjank replied:
There is no PHP standard yet. We could discuss this when such standard is being created.

Potherca replied:
Fair point. We could/should move it over when one is added.

Commit:
SQL / MySQL coding style standards

- Updates MySQL logo
- Processed feedback

Signed-off-by: Arjan Kleene <arjan.kleene@dealerdirect.nl>

mjrider commented on code-styling/mysql/README.md:
```diff
@@ -0,0 +1,574 @@
+  - [Grouping](#grouping)
+  - [Ordering](#ordering)
+  - [Limit](#limit)
+
```
I think the subject 'subqueries' is missing.
( added as duplicate, because my first omment with this text doesn't show in the conversations)

arjank replied:
I'll add something about subqueries.

mjrider commented on code-styling/mysql/README.md:
```diff
@@ -347,11 +446,12 @@ LEFT JOIN table2 AS t2
 ### Delete queries
 
-Delete queries SHOULD not be used.
+Delete queries SHOULD NOT be used. Instead a column to indicate that an item is
```
The whole previous discussion here could be summarized as 'this has nothing to do with sql styling, table design or even sql in general, but everything with the data model/storage' and thus it should not be in this document in my opinion.
The whole 'throw nothing away' gave us a big mess, lets learn from that and not document 'deleting data is bad'

Potherca replied:
@mjrider We could take it out if/when such a document is created.

Potherca commented:
All of my concerns have been addressed. Do we still need to have a discussion about backticks and/or DELETE queries? (@mjrider / @frenck )

Commit:
SQL / MySQL coding style standards

- Updates documentation about the usage of delete queries
- Adds some more bad examples based on real live queries

Signed-off-by: Arjan Kleene <arjan.kleene@dealerdirect.nl>

Potherca commented:
![Ship It Squirrel](https://dealerdirect.github.io/Memes/img/png/shipit.png)

mjrider commented on code-styling/mysql/README.md:
```diff
@@ -280,6 +292,11 @@ SELECT column_a, column_b, column_c
+-- SHA1 is calculated only once here
+SELECT *
+FROM my_table
+WHERE column_a = SHA1('some text');
```
maybe i read this wrong, but isn't this a good query ( only run sha1 once )

arjank replied:
Correct, this should be in good examples. Moved it.

Commit:
SQL / MySQL coding style standards

- Moves an example query to the correct section

Signed-off-by: Arjan Kleene <arjan.kleene@dealerdirect.nl>
